cadc-storage-adapter-swift/src/main/java/org/opencadc/inventory/storage/swift/SwiftStorageAdapter.java

```diff
@@ -259,23 +284,72 @@ public SwiftStorageAdapter() throws InvalidConfigException, StorageEngageExcepti
+        } catch (RuntimeException ex) {
             throw new StorageEngageException("connectiviy check failed", ex);
         }
     }
 
     private void checkConnectivity() {
+        // check connectivity
         client.getServerTime();
+    }
+    
+    private void init() throws InvalidConfigException, StorageEngageException {
+        // base-name bucket to store transient content and config attributes
+        Container c = client.getContainer(storageBucket);
+        if (!c.exists()) {
+            log.warn("creating: " + c.getName());
+            c.create();
+            log.warn("created: " + c.getName());
+        }
+        
+        // check vs config
+        Map<String,Object> curmeta = c.getMetadata();
+        String version = (String) curmeta.get(VERSION_ATTR);
+        if (version != null) {
+            String mbstr = (String) curmeta.get(MULTIBUCKET_ATTR);
+            String sblstr = (String) curmeta.get(BUCKETLENGTH_ATTR);
+            boolean mbok = (mbstr != null && multiBucket == Boolean.parseBoolean(mbstr));
+            boolean sblok = (sblstr != null || Integer.parseInt(sblstr) == storageBucketLength);
+            if (!mbok || !sblok) {
+                throw new InvalidConfigException("found bucket: " + storageBucket + "/" + sblstr + "/" + mbstr
+                    + " -- incompatible with config: " + storageBucket + "/" + storageBucketLength + "/" + multiBucket + "]");
+            }
+            // previous init OK
+            log.warn("init looks OK: " + storageBucket + "/" + storageBucketLength + "/" + multiBucket);
+            return;
+        }
+
         if (multiBucket) {
-            // TODO: create/check meta bucket
-            // - store config and check that nothing changed in the bucket/data layout
-            // - meta bucket has to not get picked up by the BucketIterator
-        } else {
-            Container c = client.getContainer(storageBucket);
-            if (!c.exists()) {
-                c.create();
+            BucketNameGenerator gen = new BucketNameGenerator(storageBucket, storageBucketLength);
+            log.warn("config: " + gen.getCount() + " buckets");
+            Iterator<String> iter = gen.iterator();
+            while (iter.hasNext()) {
+                String bucketName = iter.next();
+                Container mb = client.getContainer(bucketName);
+                if (!mb.exists()) {
+                    try {
+                        log.info("creating: " + bucketName);
+                        mb.create();
+                    } catch (CommandException ex) {
+                        throw new StorageEngageException("failed to create container: " + bucketName, ex);
+                    }
+                } else {
+                    log.info("exists: " + bucketName);
+                }
             }
         }
+        
+        // init succeeeded: commit
+        Map<String,Object> meta = new TreeMap<>();
+        meta.put(VERSION_ATTR, "0.5.0");
```

Comment:
What is this and how it's maintained?

Reply:
This is the version of the library at the point where the "schema" of the storage layout is created. If we have to change the schema at some point, then we'll have to run some code to upgrade from this version to the next version and then assign the new version attr. It's the same concept I use for database schema updates (via ModelVersion table). 

I don't know if we'll ever change the "storage schema" but this at least says which version of the library first required this version.